File: src/main/java/cn/chahuyun/economy/plugin/ImageManager.java
package cn.chahuyun.economy.plugin;

import cn.chahuyun.economy.BuildConstants;
import cn.chahuyun.economy.HuYanEconomy;
import cn.chahuyun.economy.utils.ImageUtil;
import cn.chahuyun.economy.utils.Log;
import lombok.Getter;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * 底图管理
 *
 * @author Moyuyanli
 * @Date 2024/8/7 21:49
 */
public class ImageManager {

    private static int next = 1;
    private static final List<BufferedImage> bufferedImages = new ArrayList<>(8);

    @Getter
    private static Font customFont;

    protected static void init(HuYanEconomy instance) throws IOException, FontFormatException {

        Log.info("开始加载字体...");

        Path path = instance.getDataFolderPath();
        File font = path.resolve("font").toFile();
        if (font.exists()) {
            File[] files = font.listFiles();
            if (files != null && files.length != 0) {
                customFont = Font.createFont(Font.TRUETYPE_FONT, files[0]).deriveFont(24f);
            }
        } else {
            customFont = new Font("宋体", Font.PLAIN, 24);
        }

        File bottom = path.resolve("bottom").toFile();
        if (!bottom.exists()) {
            return;
        }

        BufferedImage bottomPng = ImageIO.read(Objects.requireNonNull(instance.getResourceAsStream("bottom.png")));

        File[] files = bottom.listFiles();
        if (files != null) {
            int totalFiles = files.length;
            int step = totalFiles / 5 == 0 ? 1 : totalFiles / 5; // 每个阶段应该处理的文件数量
            int currentStep = 0;       // 当前处理到哪个阶段

            // 输出初始进度
            Log.info("开始加载自定义图片...");

            for (File file : files) {
                if (canBeReadAsBufferedImage(file)) {
                    try {
                        BufferedImage image = ImageIO.read(file);
                        drawBottom(image, bottomPng);
                    } catch (IOException e) {
                        Log.error("读取文件 " + file.getName() + " 出错: " + e.getMessage());
                    }
                }

                // 检查是否到达下一个阶段
                if (++currentStep % step == 0 || currentStep == totalFiles) {
                    double percentage = (double) currentStep / totalFiles * 100;
                    Log.info(String.format("处理进度: %.2f%%", percentage));
                }
            }

            // 输出完成信息
            Log.info("自定义图片和字体加载完成!");
        }
    }

    public static void view(BufferedImage image) {
        SwingUtilities.invokeLater(() -> {
            // 创建一个JFrame实例
            JFrame frame = new JFrame("Image Display");
            frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
            frame.setSize(image.getWidth() + 50, image.getHeight() + 50); // 窗口大小略大于图片

            // 创建一个 JLabel 并设置它的图标为 BufferedImage
            JLabel label = new JLabel(new ImageIcon(image));

            // 把 JLabel 添加到 JFrame 中
            frame.add(label, BorderLayout.CENTER);

            // 设置 JFrame 可见
            frame.setVisible(true);
        });
    }


    private static void drawBottom(BufferedImage background, BufferedImage bottomPng) {
        // 创建一个新的 BufferedImage
        BufferedImage canvas = new BufferedImage(1024, 576, BufferedImage.TYPE_INT_ARGB);

        // 获取 Graphics2D 对象用于绘制
        Graphics2D g2d = ImageUtil.getG2d(canvas);

        // 计算缩放比例
        double scaleWidth = (double) 1024 / background.getWidth();
        double scaleHeight = (double) 576 / background.getHeight();
        double scale = Math.max(scaleWidth, scaleHeight);

        // 缩放后的宽度和高度
        int scaledWidth = (int) (background.getWidth() * scale);
        int scaledHeight = (int) (background.getHeight() * scale);

        // 居中的位置
        int x = (1024 - scaledWidth) / 2;
        int y = (576 - scaledHeight) / 2;

        // 绘制缩放后的背景图片到新的 BufferedImage 上
        g2d.drawImage(background, x, y, scaledWidth, scaledHeight, null);

        // 绘制底
        g2d.drawImage(bottomPng, 0, 0, null);

        g2d.setFont(customFont);
        g2d.setColor(Color.BLACK);

        g2d.drawString("签到时间:", 70, 340);
        g2d.drawString("连签次数:", 70, 385);

        int infoY = 478;
        g2d.setFont(customFont.deriveFont(32f));
        g2d.drawString("我的金币", 118, infoY);
        g2d.drawString("签到获得", 358, infoY);
        g2d.drawString("我的银行", 608, infoY);
        g2d.drawString("今日收益", 858, infoY);

        g2d.setFont(customFont.deriveFont(12f));
        g2d.drawString("by Mirai & HuYanEconomy(壶言经济) v" + BuildConstants.VERSION, 730, 573);

        // 释放资源
        g2d.dispose();

        bufferedImages.add(ImageUtil.makeRoundedCorner(canvas, 30));
    }

    /**
     * 获取下一个底图
     *
     * @return 新的一个底图
     */
    public static BufferedImage getNextBottom() {
        if (bufferedImages.isEmpty()) {
            return null;
        }
        BufferedImage bufferedImage = bufferedImages.get(next++ % bufferedImages.size());
        return new BufferedImage(bufferedImage.getColorModel(),
                bufferedImage.copyData(null),
                bufferedImage.getColorModel().isAlphaPremultiplied(),
                null);
    }


    private static boolean canBeReadAsBufferedImage(File file) {
        try {
            BufferedImage image = ImageIO.read(file);
            return image != null;
        } catch (IOException e) {
            return false;
        }
    }

}

File: src/main/kotlin/cn/chahuyun/economy/plugin/ImageManager.kt
package cn.chahuyun.economy.plugin

import cn.chahuyun.economy.HuYanEconomy
import cn.chahuyun.economy.utils.EconomyImageRenderer
import cn.chahuyun.economy.utils.FormatUtil
import cn.chahuyun.economy.utils.ImageUtil
import cn.chahuyun.economy.utils.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.runBlocking
import java.awt.BorderLayout
import java.awt.Font
import java.awt.FontFormatException
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.util.concurrent.atomic.AtomicInteger
import javax.imageio.ImageIO
import javax.swing.ImageIcon
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.SwingUtilities

/**
 * 底图管理（Kotlin 实现）
 *
 * - 初始化阶段并发加载 data/bottom 下的图片并写入缓存
 * - 保持 Java 侧静态调用方式：ImageManager.init / getNextBottom / getCustomFont
 */
object ImageManager {

    @Volatile
    private var bufferedImages: List<BufferedImage> = emptyList()

    private val next = AtomicInteger(1)

    @Volatile
    private var customFont: Font = Font("宋体", Font.PLAIN, 24)

    @JvmStatic
    fun getCustomFont(): Font = customFont

    @JvmStatic
    @Throws(IOException::class, FontFormatException::class)
    fun init(instance: HuYanEconomy) {
        Log.info("开始加载字体...")

        val path = instance.dataFolderPath
        val fontDir = path.resolve("font").toFile()
        customFont = loadFontOrDefault(fontDir)

        val bottomDir = path.resolve("bottom").toFile()
        if (!bottomDir.exists()) return

        val files = bottomDir.listFiles()?.toList().orEmpty()
        val totalFiles = files.size
        if (totalFiles == 0) {
            bufferedImages = emptyList()
            Log.info("自定义图片和字体加载完成!")
            return
        }

        // 输出初始进度
        Log.info("开始加载自定义图片...")

        // 每 20% 打一次进度，保留原来的“分阶段”体验
        val step = (totalFiles / 5).let { if (it == 0) 1 else it }
        val completed = AtomicInteger(0)

        val loaded = runBlocking {
            files.map { file ->
                async(Dispatchers.IO) {
                    try {
                        loadOneBottomImage(file)
                    } finally {
                        val done = completed.incrementAndGet()
                        if (done % step == 0 || done == totalFiles) {
                            val percentage = done.toDouble() / totalFiles.toDouble() * 100.0
                            Log.info("处理进度: ${FormatUtil.fixed(percentage, 2)}%")
                        }
                    }
                }
            }.awaitAll()
        }.filterNotNull()

        bufferedImages = loaded
        Log.info("自定义图片和字体加载完成!")
    }

    @JvmStatic
    fun view(image: BufferedImage) {
        SwingUtilities.invokeLater {
            val frame = JFrame("Image Display")
            frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
            frame.setSize(image.width + 50, image.height + 50)
            frame.add(JLabel(ImageIcon(image)), BorderLayout.CENTER)
            frame.isVisible = true
        }
    }

    /**
     * 获取下一个底图（深拷贝，避免外部绘制污染缓存）
     */
    @JvmStatic
    fun getNextBottom(): BufferedImage? {
        val list = bufferedImages
        if (list.isEmpty()) return null

        val idx = (next.getAndIncrement() % list.size).let { if (it < 0) -it else it }
        val bufferedImage = list[idx]
        return BufferedImage(
            bufferedImage.colorModel,
            bufferedImage.copyData(null),
            bufferedImage.colorModel.isAlphaPremultiplied,
            null
        )
    }

    private fun loadFontOrDefault(fontDir: File): Font {
        if (!fontDir.exists()) {
            return Font("宋体", Font.PLAIN, 24)
        }
        val files = fontDir.listFiles()
        if (files.isNullOrEmpty()) {
            return Font("宋体", Font.PLAIN, 24)
        }
        // 兼容旧逻辑：取目录下第一个字体文件
        val fontFile = files[0]
        return try {
            Font.createFont(Font.TRUETYPE_FONT, fontFile).deriveFont(24f)
        } catch (_: Exception) {
            Font("宋体", Font.PLAIN, 24)
        }
    }

    private fun loadOneBottomImage(file: File): BufferedImage? {
        if (!file.isFile) return null
        val background = try {
            ImageIO.read(file)
        } catch (e: IOException) {
            Log.error("读取文件 ${file.name} 出错: ${e.message}")
            null
        }
        if (background == null) return null
        return drawBottom(background)
    }

    /**
     * 生成一个“已合成底 + 圆角”的底图缓存项
     */
    private fun drawBottom(background: BufferedImage): BufferedImage {
        val canvas = BufferedImage(EconomyImageRenderer.PERSONAL_WIDTH, EconomyImageRenderer.PERSONAL_HEIGHT, BufferedImage.TYPE_INT_ARGB)
        val g2d: Graphics2D = ImageUtil.getG2d(canvas)

        val scaleWidth = EconomyImageRenderer.PERSONAL_WIDTH.toDouble() / background.width.toDouble()
        val scaleHeight = EconomyImageRenderer.PERSONAL_HEIGHT.toDouble() / background.height.toDouble()
        val scale = maxOf(scaleWidth, scaleHeight)

        val scaledWidth = (background.width * scale).toInt()
        val scaledHeight = (background.height * scale).toInt()

        val x = (EconomyImageRenderer.PERSONAL_WIDTH - scaledWidth) / 2
        val y = (EconomyImageRenderer.PERSONAL_HEIGHT - scaledHeight) / 2

        g2d.drawImage(background, x, y, scaledWidth, scaledHeight, null)

        g2d.dispose()
        return ImageUtil.makeRoundedCorner(canvas, 30)
    }
}

